internal/datastore/crdb/reader.go

```diff
@@ -0,0 +1,240 @@
+package crdb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"github.com/jackc/pgx/v4"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	"github.com/authzed/spicedb/internal/datastore/options"
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+const (
+	errUnableToReadConfig     = "unable to read namespace config: %w"
+	errUnableToListNamespaces = "unable to list namespaces: %w"
+)
+
+var (
+	queryReadNamespace = psql.Select(colConfig, colTimestamp).From(tableNamespace)
+
+	queryTuples = psql.Select(
+		colNamespace,
+		colObjectID,
+		colRelation,
+		colUsersetNamespace,
+		colUsersetObjectID,
+		colUsersetRelation,
+	).From(tableTuple)
+
+	schema = common.SchemaInformation{
+		ColNamespace:        colNamespace,
+		ColObjectID:         colObjectID,
+		ColRelation:         colRelation,
+		ColUsersetNamespace: colUsersetNamespace,
+		ColUsersetObjectID:  colUsersetObjectID,
+		ColUsersetRelation:  colUsersetRelation,
+	}
+)
+
+type crdbReader struct {
+	txSource      common.TransactionFactory
+	querySplitter common.TupleQuerySplitter
+	keyer         overlapKeyer
+	overlapKeySet keySet
+	execute       executeTxRetryFunc
+}
+
+func (cr *crdbReader) ReadNamespace(
+	ctx context.Context,
+	nsName string,
+) (*core.NamespaceDefinition, datastore.Revision, error) {
+	ctx = datastore.SeparateContextWithTracing(ctx)
```

Comment:
It would be interesting to know when this is even required. We added this because any time we canceled the context upstream, it would cancel the context downstream and kill the PG connection. I'm not even sure if this is required for some of the datastores...